engine.go

```diff
@@ -261,12 +227,47 @@ func (e *EngineBackend) ExecutePayload(ctx context.Context, payload *ExecutionPa
 	return &ExecutePayloadResult{Status: ExecutionValid}, nil
 }
 
-func (e *EngineBackend) ConsensusValidated(ctx context.Context, params *ConsensusValidatedParams) error {
-	e.log.WithField("params", params).Info("Consensus validated")
-	return nil
-}
+func (e *EngineBackend) ForkchoiceUpdatedV1(ctx context.Context, heads *ForkchoiceState, attributes *PayloadAttributes) (*ForkchoiceUpdatedResult, error) {
+	e.log.WithField("forkchoice state", heads).Info("Forkchoice validated")
 
-func (e *EngineBackend) ForkchoiceUpdated(ctx context.Context, params *ForkchoiceUpdatedParams) error {
-	e.log.WithField("params", params).Info("Forkchoice validated")
-	return nil
+	if attributes == nil {
+		return nil, nil
+	}
+	id := atomic.AddUint64(&e.payloadIdCounter, 1)
+
+	plog := e.log.WithField("payload_id", id)
+	plog.WithField("attributes", attributes).Info("Preparing new payload")
+
+	gasLimit := e.mockChain.gspec.GasLimit
+	txsCreator := TransactionsCreator(func(config *params.ChainConfig, bc core.ChainContext,
+		statedb *state.StateDB, header *types.Header, cfg vm.Config) []*types.Transaction {
+		// empty payload
+
+		// TODO: maybe vary these a little?
+		return nil
+	})
+	extraData := []byte{}
+
+	bl, err := e.mockChain.AddNewBlock(common.BytesToHash(heads.HeadBlockHash[:]), attributes.FeeRecipient, uint64(attributes.Timestamp),
+		gasLimit, txsCreator, extraData, nil, false)
+
+	if err != nil {
+		// TODO: proper error codes
+		plog.WithError(err).Error("Failed to create block, cannot build new payload")
+		return nil, err
+	}
+
+	payload, err := BlockToPayload(bl, attributes.Random)
+	if err != nil {
+		plog.WithError(err).Error("Failed to convert block to payload")
+		// TODO: proper error codes
+		return nil, err
+	}
+
+	// store in cache for later retrieval
+	e.recentPayloads.Add(id, payload)
+
+	buf := make(hexutil.Bytes, 8)
+	binary.PutUvarint(buf, id)
```

Comment:
This should be `binary.LittleEndian.PutUint64` (or big endian). With 8 bytes id and uint64 internally we don't need var-ints.

Reply:
Oh, and due to typing the payload cache might be broken.